mos-platform/common/c/stdio-full.c

```diff
@@ -87,6 +85,86 @@ void _stdio_closeall(void) {
     fclose(f);
 }
 
+/* A system call that writes a stream's buffer.
+   Returns 0 on success, EOF on write error.
+   Sets stream error flags and errno appropriately on error.
+*/
+static int flush_buffer(FILE *stream) {
+  // No need to handle buffers > INT_MAX, as PDCLib doesn't allow them */
+  size_t written = 0;
+  int rc;
+
+  // Note: everything written to the buffer has already been converted.
+  // Otherwise, when write errors occur, it's unclear what still remains to be
+  // converted.
+
+  // Keep trying to write data until everything is written or an error occurs.
+  for (;;) {
+    rc = write(stream->handle, stream->buffer + written,
+               stream->bufidx - written);
+
+    if (rc < 0) {
+      /* Flag the stream */
+      stream->status |= ERRORFLAG;
+      /* Move unwritten remains to begin of buffer. */
+      stream->bufidx -= written;
+      memmove(stream->buffer, stream->buffer + written, stream->bufidx);
+      return EOF;
+    }
+
+    written += (size_t)rc;
+    stream->pos += rc;
+
+    if (written == stream->bufidx) {
+      /* Buffer written completely. */
+      stream->bufidx = 0;
+      return 0;
+    }
+  }
+}
+
+static int stdio_close(FILE *stream) {
+  int rc = 0;
+
+  /* Flush buffer */
+  if (stream->status & FWRITE)
+    rc = flush_buffer(stream);
+
+  /* Close handle */
+  if (close(stream->handle) != 0)
+    rc = EOF;
+
+  /* Delete tmpfile() */
+  if (stream->del_filename) {
+    if (remove(stream->del_filename) != 0)
+      rc = EOF;
+    free(stream->del_filename);
+  }
+  return rc;
+}
+
+static void free_stream(FILE *stream) {
+  /* Remove stream from list */
+  if (filelist == stream)
```

Comment:
nit: One aspect of the LLVM style I'd like to keep is that if one clause of an if/else has braces; the other must also. Making them the two clauses visually similar tends to help with finding their ranges at a glance.

Reply:
Done.